File: src/Framework/Actors/ActorCollection.fs
namespace Animations.Actors

open Animations
open PropertyHelpers

type internal ActorCollection(properties: AnimationProperties, actors: seq<IActor * AnimationTimePoint * AnimationTimePoint>) = 
    
    let actors = 
        actors
        |> Seq.map (fun (actor, start, finish) -> actor, GetFrameNumber(start, properties), GetFrameNumber(finish, properties))
        |> List.ofSeq

    member public this.InvokeActors(context: DrawContext) = 
        
        for (actor, start, finish) in actors do

            if ((start <= context.FrameNumber) && (context.FrameNumber < finish)) then
                actor.Perform(context)

File: Animations.Framework/ActorCollection.cs
using System.Collections.Generic;
using System.Linq;

namespace Animations.Framework
{
    internal class ActorCollection
    {
        private struct ValueType
        {
            public IActor Actor;
            public int Start;
            public int Finish;
        }

        private readonly ValueType[] _actors;

        public ActorCollection(AnimationProperties properties, IEnumerable<ActorEntry> actors)
        {
            _actors = actors.Select(tuple => new ValueType
            {
                Actor = tuple.Actor,
                Start = tuple.Start.GetFrameNumber(properties),
                Finish = tuple.Finish.GetFrameNumber(properties)
            }).ToArray();
        }

        public void InvokeActors(IDrawContext context)
        {
            foreach (var item in _actors)
            {
                if ((item.Start <= context.FrameNumber) && (context.FrameNumber < item.Finish))
                {
                    item.Actor.Perform(context);
                }
            }
        }
    }
}